### PR TITLE
Updated and Added Dictionary Commands

### PR DESCRIPTION
**Pull request by openbots-aw**
- Updated dictionary commands
- Added new dictionary commands (GetKeyFromKeyValuePairCommand, GetValueFromKeyValuePairCommand)
- Updated loop collection command to loop through dictionaries
- Added support for dictionary value types (Datatable, MailItem, MimeMessage, IWebElement)
- Added UpdateDictionaryValueCommand
- Added RemoveDictionaryValueCommand
- Added RemoveDictionaryKeyValuePairCommand
- Added GetDictionaryCountCommand
- Updated OpenBots.Core to support dictionary changes

### CODE REVIEW

File: OpenBots.Commands/OpenBots.Commands.Dictionary/CreateDictionaryCommand.cs [hunk 1]
```diff
@@ -50,28 +65,92 @@ public CreateDictionaryCommand()
...
+				case "DataTable":
+					outputDictionary = new Dictionary<string, DataTable>();
+						foreach (DataRow rwColumnName in v_ColumnNameDataTable.Rows)
+						{
+							DataTable dataTable;
+							var dataTableVariable = rwColumnName.Field<string>("Values").ConvertUserVariableToObject(engine);
+							if (dataTableVariable != null && dataTableVariable is DataTable)
+								dataTable = (DataTable)dataTableVariable;
+							else
+								throw new Exception("Invalid dictionary value type, please provide valid dictionary value type.");
```
Comment (openbots-nc) [inline OpenBots.Commands/OpenBots.Commands.Dictionary/CreateDictionaryCommand.cs:97]: DataException

File: OpenBots.Commands/OpenBots.Commands.Dictionary/CreateDictionaryCommand.cs [hunk 2]
```diff
@@ -50,28 +65,92 @@ public CreateDictionaryCommand()
...
+				case "MailItem (Outlook)":
+					outputDictionary = new Dictionary<string, MailItem>();
+					foreach (DataRow rwColumnName in v_ColumnNameDataTable.Rows)
+						{
+							MailItem mailItem;
+							var mailItemVariable = rwColumnName.Field<string>("Values").ConvertUserVariableToObject(engine);
+							if (mailItemVariable != null && mailItemVariable is MailItem)
+								mailItem = (MailItem)mailItemVariable;
+							else
+								throw new Exception("Invalid dictionary value type, please provide valid dictionary value type.");
```
Comment (openbots-nc) [inline OpenBots.Commands/OpenBots.Commands.Dictionary/CreateDictionaryCommand.cs:111]: DataException

File: OpenBots.Commands/OpenBots.Commands.Dictionary/CreateDictionaryCommand.cs [hunk 3]
```diff
@@ -50,28 +65,92 @@ public CreateDictionaryCommand()
...
+				case "MimeMessage (IMAP/SMTP)":
+					outputDictionary = new Dictionary<string, MimeMessage>();
+					foreach (DataRow rwColumnName in v_ColumnNameDataTable.Rows)
+						{
+							MimeMessage mimeMessage;
+							var mimeMessageVariable = rwColumnName.Field<string>("Values").ConvertUserVariableToObject(engine);
+							if (mimeMessageVariable != null && mimeMessageVariable is MimeMessage)
+								mimeMessage = (MimeMessage)mimeMessageVariable;
+							else
+							throw new Exception("Invalid dictionary value type, please provide valid dictionary value type.");
```
Comment (openbots-nc) [inline OpenBots.Commands/OpenBots.Commands.Dictionary/CreateDictionaryCommand.cs:125]: DataException

File: OpenBots.Commands/OpenBots.Commands.Dictionary/CreateDictionaryCommand.cs [hunk 4]
```diff
@@ -50,28 +65,92 @@ public CreateDictionaryCommand()
...
+				case "IWebElement":
+					outputDictionary = new Dictionary<string, IWebElement>();
+					foreach (DataRow rwColumnName in v_ColumnNameDataTable.Rows)
+						{
+							IWebElement webElement;
+							var webElementVariable = rwColumnName.Field<string>("Values").ConvertUserVariableToObject(engine);
+							if (webElementVariable != null && webElementVariable is IWebElement)
+								webElement = (IWebElement)webElementVariable;
+							else
+								throw new Exception("Invalid dictionary value type, please provide valid dictionary value type.");
```
Comment (openbots-nc) [inline OpenBots.Commands/OpenBots.Commands.Dictionary/CreateDictionaryCommand.cs:139]: DataException

File: OpenBots.Commands/OpenBots.Commands.Dictionary/GetDictionaryValueCommand.cs [hunk 5]
```diff
@@ -57,17 +62,23 @@ public override void RunCommand(object sender)
...
+			else if (v_InputDictionary.ConvertUserVariableToObject(engine) is Dictionary<string, DataTable>)
+				dict = (Dictionary<string, DataTable>)v_InputDictionary.ConvertUserVariableToObject(engine);
+			else if (v_InputDictionary.ConvertUserVariableToObject(engine) is Dictionary<string, MailItem>)
+				dict = (Dictionary<string, MailItem>)v_InputDictionary.ConvertUserVariableToObject(engine);
+			else if (v_InputDictionary.ConvertUserVariableToObject(engine) is Dictionary<string, MimeMessage>)
+				dict = (Dictionary<string, MimeMessage>)v_InputDictionary.ConvertUserVariableToObject(engine);
+			else if (v_InputDictionary.ConvertUserVariableToObject(engine) is Dictionary<string, IWebElement>)
+				dict = (Dictionary<string, IWebElement>)v_InputDictionary.ConvertUserVariableToObject(engine);
+			else
+				throw new Exception("Invalid dictionary value type, please provide valid dictionary value type.");
```
Comment (openbots-nc) [inline OpenBots.Commands/OpenBots.Commands.Dictionary/GetDictionaryValueCommand.cs:79]: DataException

File: OpenBots.Commands/OpenBots.Commands.Dictionary/GetKeyFromKeyValuePairCommand .cs [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+			else if (v_InputKeyValuePair.ConvertUserVariableToObject(engine) is KeyValuePair<string, DataTable>)
+				keyValuePair = (KeyValuePair<string, DataTable>)v_InputKeyValuePair.ConvertUserVariableToObject(engine);
+			else if (v_InputKeyValuePair.ConvertUserVariableToObject(engine) is KeyValuePair<string, MailItem>)
+				keyValuePair = (KeyValuePair<string, MailItem>)v_InputKeyValuePair.ConvertUserVariableToObject(engine);
+			else if (v_InputKeyValuePair.ConvertUserVariableToObject(engine) is KeyValuePair<string, MimeMessage>)
+				keyValuePair = (KeyValuePair<string, MimeMessage>)v_InputKeyValuePair.ConvertUserVariableToObject(engine);
+			else if (v_InputKeyValuePair.ConvertUserVariableToObject(engine) is KeyValuePair<string, IWebElement>)
+				keyValuePair = (KeyValuePair<string, IWebElement>)v_InputKeyValuePair.ConvertUserVariableToObject(engine);
+			else
+				throw new Exception("Invalid dictionary value type, please provide valid dictionary value type.");
```
Comment (openbots-nc) [inline OpenBots.Commands/OpenBots.Commands.Dictionary/GetKeyFromKeyValuePairCommand .cs:66]: DataException

File: OpenBots.Commands/OpenBots.Commands.Dictionary/GetValueFromKeyValuePairCommand.cs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+			else if (v_InputKeyValuePair.ConvertUserVariableToObject(engine) is KeyValuePair<string, DataTable>)
+				keyValuePair = (KeyValuePair<string, DataTable>)v_InputKeyValuePair.ConvertUserVariableToObject(engine);
+			else if (v_InputKeyValuePair.ConvertUserVariableToObject(engine) is KeyValuePair<string, MailItem>)
+				keyValuePair = (KeyValuePair<string, MailItem>)v_InputKeyValuePair.ConvertUserVariableToObject(engine);
+			else if (v_InputKeyValuePair.ConvertUserVariableToObject(engine) is KeyValuePair<string, MimeMessage>)
+				keyValuePair = (KeyValuePair<string, MimeMessage>)v_InputKeyValuePair.ConvertUserVariableToObject(engine);
+			else if (v_InputKeyValuePair.ConvertUserVariableToObject(engine) is KeyValuePair<string, IWebElement>)
+				keyValuePair = (KeyValuePair<string, IWebElement>)v_InputKeyValuePair.ConvertUserVariableToObject(engine);
+			else
+				throw new Exception("Invalid dictionary value type, please provide valid dictionary value type.");
```
Comment (openbots-nc) [inline OpenBots.Commands/OpenBots.Commands.Dictionary/GetValueFromKeyValuePairCommand.cs:66]: DataException

File: OpenBots.Commands/OpenBots.Commands.Loop/LoopCollectionCommand.cs [hunk 8]
```diff
@@ -87,6 +88,26 @@ public override void RunCommand(object sender, ScriptAction parentCommand)
 			{
 				listToLoop = (List<MimeMessage>)complexVariable;
 			}
+			else if (complexVariable is Dictionary<string, string>)
+			{
+                listToLoop = ((Dictionary<string, string>)complexVariable).ToList();
+            }
```
Comment (openbots-nc) [inline OpenBots.Commands/OpenBots.Commands.Loop/LoopCollectionCommand.cs:94]: Fix indentation here

File: OpenBots.Core/Utilities/CommonUtilities/VariableMethods.cs [hunk 9]
```diff
@@ -122,6 +122,22 @@ public static string ConvertUserVariableToString(this string userInputString, IA
...
+
+                            dynamic pair;
+                            string resultItem;
+                            pair = varCheck.VariableValue;
+                            if (resultType.ToLower() == "key")
+                                resultItem = pair.Key;
+                            else if(resultType.ToLower() == "value")
+                                resultItem = StringMethods.ConvertObjectToString(pair.Value);
+                            else
+                                throw new Exception("Only use of key and value is allowed using dot operater with keyvaluepair is allowed");
```
Comment (openbots-nc) [inline OpenBots.Core/Utilities/CommonUtilities/VariableMethods.cs:138]: DataException

**Review by openbots-nc — CHANGES_REQUESTED** [review]

File: OpenBots.Commands/OpenBots.Commands.Dictionary/GetKeyFromKeyValuePairCommand .cs [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+		[DisplayName("KeyValuePair")]
+		[Description("Specify the KeyValuePair variable to get a key from.")]
+		[SampleUsage("{vKeyValuePair}")]
+		[Remarks("")]
+		[Editor("ShowVariableHelper", typeof(UIAdditionalHelperType))]
+		public string v_InputKeyValuePair { get; set; }
+
+		[Required]
+		[Editable(false)]
+		[DisplayName("Output Value Variable")]
```
Comment (openbots-ff) [inline OpenBots.Commands/OpenBots.Commands.Dictionary/GetKeyFromKeyValuePairCommand .cs:35]:
```suggestion
		[DisplayName("Output Key Variable")]
```

File: OpenBots.Core/Utilities/CommonUtilities/StringMethods.cs [hunk 11]
```diff
@@ -160,6 +162,12 @@ public static string ConvertObjectToString(object obj)
                     case string a when a.Contains("System.Collections.Generic.Dictionary`2[[System.String") && a.Contains("],[System.String"):
                     case string b when b.Contains("System.Collections.Generic.Dictionary`2[[System.String") && b.Contains("],[System.Object"):
```
Comment (openbots-ff) [inline OpenBots.Core/Utilities/CommonUtilities/StringMethods.cs:163]: It looks like you kept a generic object for dictionaries here but you were more specific with the keyvalue pairs. Did you test whether they both render properly in the debug pane/show message window? Please keep in mind that this Dictionary<string, object> type was created for the WorkQueueItem command and that needs to still render correctly after all these changes.

**Review by openbots-ff — CHANGES_REQUESTED** [review]

**Review by openbots-ab — COMMENTED** [review]
Please add the tags for the new commands in .nuspec file of this project.

File: OpenBots.Commands/OpenBots.Commands.Dictionary/AddDictionaryItemCommand.cs [hunk 12]
```diff
@@ -56,16 +60,84 @@ public override void RunCommand(object sender)
...
+							webElement = (IWebElement)webElementVariable;
+						else
+							throw new Exception("Invalid dictionary value type, please provide valid dictionary value type.");
+						((Dictionary<string, IWebElement>)dictionaryVariable).Add(
+							rwColumnName.Field<string>("Keys").ConvertUserVariableToString(engine), webElement);
+					}
+				}
+				else
+				{
+					throw new Exception("Dictionary type Not Supported");
```
[the remark this answers is not in the document]
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.Dictionary/AddDictionaryItemCommand.cs]: Done.

File: OpenBots.Commands/OpenBots.Commands.Dictionary/AddDictionaryItemCommand.cs [hunk 13]
```diff
@@ -56,16 +60,84 @@ public override void RunCommand(object sender)
...
+				else if (dictionaryVariable is Dictionary<string, MailItem>)
+				{
+					foreach (DataRow rwColumnName in v_ColumnNameDataTable.Rows)
+					{
+						MailItem mailItem;
+						var mailItemVariable = rwColumnName.Field<string>("Values").ConvertUserVariableToObject(engine);
+						if (mailItemVariable != null && mailItemVariable is MailItem)
+							mailItem = (MailItem)mailItemVariable;
+						else
+							throw new Exception("Invalid dictionary value type, please provide valid dictionary value type.");
```
[the remark this answers is not in the document]
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.Dictionary/AddDictionaryItemCommand.cs]: Done.

File: OpenBots.Commands/OpenBots.Commands.Dictionary/AddDictionaryItemCommand.cs [hunk 14]
```diff
@@ -56,16 +60,84 @@ public override void RunCommand(object sender)
...
+				else if (dictionaryVariable is Dictionary<string, MimeMessage>)
+				{
+					foreach (DataRow rwColumnName in v_ColumnNameDataTable.Rows)
+					{
+						MimeMessage mimeMessage;
+						var mimeMessageVariable = rwColumnName.Field<string>("Values").ConvertUserVariableToObject(engine);
+						if (mimeMessageVariable != null && mimeMessageVariable is MimeMessage)
+							mimeMessage = (MimeMessage)mimeMessageVariable;
+						else
+							throw new Exception("Invalid dictionary value type, please provide valid dictionary value type.");
```
[the remark this answers is not in the document]
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.Dictionary/AddDictionaryItemCommand.cs]: Done.

File: OpenBots.Commands/OpenBots.Commands.Dictionary/AddDictionaryItemCommand.cs [hunk 15]
```diff
@@ -56,16 +60,84 @@ public override void RunCommand(object sender)
...
+				else if (dictionaryVariable is Dictionary<string, IWebElement>)
+				{
+					foreach (DataRow rwColumnName in v_ColumnNameDataTable.Rows)
+					{
+						IWebElement webElement;
+						var webElementVariable = rwColumnName.Field<string>("Values").ConvertUserVariableToObject(engine);
+						if (webElementVariable != null && webElementVariable is IWebElement)
+							webElement = (IWebElement)webElementVariable;
+						else
+							throw new Exception("Invalid dictionary value type, please provide valid dictionary value type.");
```
[the remark this answers is not in the document]
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.Dictionary/AddDictionaryItemCommand.cs]: Done.

File: OpenBots.Commands/OpenBots.Commands.Dictionary/AddDictionaryItemCommand.cs [hunk 16]
```diff
@@ -56,16 +60,84 @@ public override void RunCommand(object sender)
...
-
-			foreach (DataRow rwColumnName in v_ColumnNameDataTable.Rows)
+			((object)dictionaryVariable).StoreInUserVariable(engine, v_DictionaryName);
+			}
+			else
 			{
-				outputDictionary.Add(
-					rwColumnName.Field<string>("Keys").ConvertUserVariableToString(engine), 
-					rwColumnName.Field<string>("Values").ConvertUserVariableToString(engine));
+				throw new Exception("Attempted to add data to a variable, but the variable was not found. Enclose variables within braces, ex. {vVariable}");
```
[the remark this answers is not in the document]
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.Dictionary/AddDictionaryItemCommand.cs]: Done.

File: OpenBots.Commands/OpenBots.Commands.Dictionary/GetDictionaryValueCommand.cs (hunk 5 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.Dictionary/GetDictionaryValueCommand.cs]: Done.

File: OpenBots.Commands/OpenBots.Commands.Dictionary/GetKeyFromKeyValuePairCommand .cs (hunk 6 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.Dictionary/GetKeyFromKeyValuePairCommand .cs]: Done.

File: OpenBots.Commands/OpenBots.Commands.Dictionary/GetValueFromKeyValuePairCommand.cs (hunk 7 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.Dictionary/GetValueFromKeyValuePairCommand.cs]: Done.

File: OpenBots.Commands/OpenBots.Commands.Loop/LoopCollectionCommand.cs (hunk 8 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.Loop/LoopCollectionCommand.cs:93]: Done.

File: OpenBots.Core/Utilities/CommonUtilities/VariableMethods.cs (hunk 9 above)
Comment (openbots-aw) [reply OpenBots.Core/Utilities/CommonUtilities/VariableMethods.cs]: Done.

File: OpenBots.Commands/OpenBots.Commands.Dictionary/GetKeyFromKeyValuePairCommand .cs (hunk 10 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.Dictionary/GetKeyFromKeyValuePairCommand .cs]: Done.

File: OpenBots.Core/Utilities/CommonUtilities/StringMethods.cs (hunk 11 above)
Comment (openbots-aw) [reply OpenBots.Core/Utilities/CommonUtilities/StringMethods.cs]: I have added all supported types, and at the end I have included the Dictionary<string, object> in StringMethods.cs. I have tested rendering in debug pane/show message for all dictionary types.

File: OpenBots.Commands/OpenBots.Commands.Dictionary/OpenBots.Commands.Dictionary.nuspec [hunk 17]
```diff
@@ -14,7 +14,7 @@
     <description>OpenBots commands for creating and manipulating Dictionary objects.</description>
     <releaseNotes>https://openbots.ai/category/release-notes/</releaseNotes>
     <copyright>(C) OpenBots Inc 2020-2021</copyright>
-    <tags>OpenBots Studio Dictionary Commands AddDictionaryItem CreateDictionary GetDictionaryValue</tags>
+    <tags>OpenBots Studio Dictionary Commands AddDictionaryItem CreateDictionary GetDictionaryValue GetKeyFromKeyValuePairCommand GetValueFromKeyValuePairCommand</tags>
```
Comment (openbots-ff) [inline OpenBots.Commands/OpenBots.Commands.Dictionary/OpenBots.Commands.Dictionary.nuspec:17]:
```suggestion
    <tags>OpenBots Studio Dictionary Commands AddDictionaryItem CreateDictionary GetDictionaryValue GetKeyFromKeyValuePair GetValueFromKeyValuePair</tags>
```

**Review by openbots-ff — CHANGES_REQUESTED** [review]

File: OpenBots.Commands/OpenBots.Commands.Dictionary/UpdateDictionaryValueCommand.cs [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+		[Editor("ShowVariableHelper", typeof(UIAdditionalHelperType))]
+		public string v_Value { get; set; }
+
+		[Required]
+		[DisplayName("Key")]
+		[Description("Enter the Key where the value will be updated.")]
+		[SampleUsage("Hello || {vKey}")]
+		[Remarks("Providing a non existing key will produce an exception.")]
+		[Editor("ShowVariableHelper", typeof(UIAdditionalHelperType))]
+		public string v_Key { get; set; }
```
Comment (openbots-ff) [inline OpenBots.Commands/OpenBots.Commands.Dictionary/UpdateDictionaryValueCommand.cs:47]: Could you swap the order of key and value here, and in the render method?

**Review by openbots-ff — CHANGES_REQUESTED** [review]

File: OpenBots.Commands/OpenBots.Commands.Dictionary/RemoveDictionaryValueCommand.cs [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+
+			RenderedControls.AddRange(commandControls.CreateDefaultInputGroupFor("v_DictionaryName", this, editor));
+			RenderedControls.AddRange(commandControls.CreateDefaultInputGroupFor("v_Key", this, editor));
+
+			return RenderedControls;
+		}
+
+		public override string GetDisplayValue()
+		{
+			return base.GetDisplayValue() + $" [Remove Value from Dictionary '{v_DictionaryName}' at Value '{v_Key}']";
```
Comment (openbots-ff) [inline OpenBots.Commands/OpenBots.Commands.Dictionary/RemoveDictionaryValueCommand.cs:87]:
```suggestion
			return base.GetDisplayValue() + $" [Remove Value from Dictionary '{v_DictionaryName}' at Key '{v_Key}']";
```

**Review by openbots-ff — CHANGES_REQUESTED** [review]

File: OpenBots.Commands/OpenBots.Commands.Dictionary/UpdateDictionaryValueCommand.cs (hunk 18 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.Dictionary/UpdateDictionaryValueCommand.cs:39]: Done.

**Review by openbots-ff — APPROVED** [review]

File: OpenBots.Commands/OpenBots.Commands.Dictionary/CreateDictionaryCommand.cs (hunk 1 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.Dictionary/CreateDictionaryCommand.cs]: Done.

File: OpenBots.Commands/OpenBots.Commands.Dictionary/CreateDictionaryCommand.cs (hunk 2 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.Dictionary/CreateDictionaryCommand.cs]: Done

File: OpenBots.Commands/OpenBots.Commands.Dictionary/CreateDictionaryCommand.cs (hunk 3 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.Dictionary/CreateDictionaryCommand.cs]: Done.

File: OpenBots.Commands/OpenBots.Commands.Dictionary/CreateDictionaryCommand.cs (hunk 4 above)
Comment (openbots-aw) [reply OpenBots.Commands/OpenBots.Commands.Dictionary/CreateDictionaryCommand.cs]: Done.

**Review by openbots-ff — APPROVED** [review]

**Review by openbots-ab — APPROVED** [review]

**Review by openbots-ch — APPROVED** [review]

**Review by openbots-nc — APPROVED** [review]

**Pull request merged**